mlir/lib/Dialect/Rock/Transforms/RockPipeline.cpp

```
@@ -0,0 +1,492 @@
+  auto maybeNumIterations =
+      rock::computeConstDiff(forOp.getLowerBound(), forOp.getUpperBound());
+  if (!maybeNumIterations.has_value() ||
+      (maybeNumIterations.has_value() && maybeNumIterations.value() > 1)) {
+    const bool placeBarrier = true;
+    const bool after = false;
+    extendedStages.push_back(
+        placeEmptyStage(rewriter, loc, stages.back(), placeBarrier, after));
+  }
+  initiationInterval *= 2;
+}
+
+struct RockPipeline : public rock::impl::RockPipelinePassBase<RockPipeline> {
+  using rock::impl::RockPipelinePassBase<RockPipeline>::RockPipelinePassBase;
+  void runOnOperation() override;
+};
+
+} // end namespace
+
+void RockPipeline::runOnOperation() {
+  func::FuncOp func = getOperation();
+  MLIRContext *ctx = func->getContext();
+  Location loc = func->getLoc();
+  IRRewriter rewriter(ctx);
+
+  std::map<Operation *, SmallVector<rock::GpuAllocOp>> ldsAllocMap;
+
+  llvm::DenseSet<rock::GpuAllocOp> allocs;
+
+  // Collect the global resources (i.e., the memory allocations)
+  // Note: we can only have two kind of memory:
+  // - Registers
+  // - LDS
+  func.walk([&](rock::GpuAllocOp alloc) { allocs.insert(alloc); });
+
+  DenseMap<rock::GpuAllocOp, int> multiBufferFactors;
+  llvm::DenseMap<scf::ForOp, ScheduleType> scheduleMap;
+  llvm::DenseMap<scf::ForOp, Operation *> nextOpMap;
+  for (auto res : allocs)
+    multiBufferFactors[res] = 1;
+
+  auto rockPipelineAttrName = rock::PipelineAttr::getMnemonic();
+  func.walk([&](scf::ForOp forOp) -> WalkResult {
+    SmallVector<rock::StageOp> stages;
+
+    if (!forOp->hasAttrOfType<rock::PipelineAttr>(rockPipelineAttrName))
+      return WalkResult::advance();
+
+    bool isNestedPipelining = false;
+    forOp.getBody()->walk([&](scf::ForOp nestedFor) {
+      if (nestedFor->hasAttr(rockPipelineAttrName))
+        isNestedPipelining = true;
+    });
+
+    if (isNestedPipelining)
+      return WalkResult::advance();
```

I think we should error out for now rather than silently ignoring.
So the user would not be mislead to believe the pipelining semantics the user provided are legit.